plotly/tools.py

```diff
@@ -1269,3 +1276,547 @@ def return_figure_from_figure_or_data(figure_or_data, validate_figure):
             )
 
     return figure
+
+
+def Quiver(x, y, u, v,
+           scale=.1, angle=np.pi/9,
+           arrow_scale=.3, barb_color='rgb(114, 132, 314)',
+           arrow_color='rgb(114, 132, 314)', arrow_width=1,
+           barb_width=1, **kwargs):
+    """Return a data object for a quiver plot.
+    x, y, u, and v can be np.ndarrays of equal dimmensions,
+    or lists where x & y are the same length and u & v are the same length.
+    x and y are the coordinates of the arrow locations.
+    u and v give the x and y components of the arrow vectors.
+
+    Example 1:
+    ```
+    # Just 1 Arrow from (0,0) to (1,1)
+    data = Quiver(x=[0], y=[0], u=[1], v=[1], scale=1)
+    ```
+
+    Example 2:
+    ```
+    x,y = np.meshgrid(np.arange(0,np.pi,.2),np.arange(0,np.pi,.2) )
+    u = np.cos(x)
+    v = np.sin(y)
+    data = Quiver(x, y, u, v)
+    ```
+
+    Example 3:
+    ```
+    x,y = np.meshgrid(np.arange(0,2,.2),np.arange(0,2,.2))
+    u = np.cos(x)*y;
+    v = np.sin(x)*y;
+    data = Quiver(x, y, u, v)
+    ```
+
+    Keywords arguments with constant defaults:
+
+    scale (kwarg, float in [0,1], default = .1):
+        Scales size of the arrows (ideally to avoid overlap).
+
+    angle (kwarg, angle in radians, default = np.pi/9):
+        Angle of arrowhead.
+
+    arrow_scale (kwarg, float in [0,1], default = 0.3):
+        Value multiplied to length of barb to get length of arrowhead.
+
+    barb_color (kwarg, color string, default = color):
+        Set color of barbs.
+
+    arrow_color (kwarg, color string, default = color):
+        Set color of arrow.
+
+    barb_width (kwarg, int greater than or equal to 1, default = width):
+        Change width of lines for barbs.
+
+    arrow_width (kwarg, int greater than or equal to 1, default = width):
+        Change width of lines for arrows.
+    """
+
+    if _numpy_imported is False:
+        raise Exception("To use Quiver() please import numpy as np")
+
+    if len(x) != len(y):
+        raise Exception("x, y, u, and v should be ndarrays of equal dimmension"
+                        "or lists where x and y are the same length"
+                        "and u and v are the same length")
+
+    if len(u) != len(v):
+        raise Exception("x, y, u, and v should be ndarrays of equal dimmension"
+                        "or lists where x and y are the same length"
+                        "and u and v are the same length")
+
+    VALID_KWARGS = ['angle', 'scale',
+                    'arrow_scale', 'barb_color',
+                    'arrow_color', 'barb_width',
+                    'arrow_width']
+    for key in kwargs.keys():
+        if key not in VALID_KWARGS:
+            raise Exception("Invalid keyword argument: '{0}'".format(key))
+
+    # Make x y u v into np.array in case user entered list
+    x = np.array(x)
+    y = np.array(y)
+    u = np.array(u)
+    v = np.array(v)
+
+    # Make arrays of x & y start values
+    Xstart = x.flatten('F')
+    Ystart = y.flatten('F')
```

Comment:
also, the alternative to flatten is to use something like this:

http://stackoverflow.com/questions/952914/making-a-flat-list-out-of-list-of-lists-in-python

```
[item for sublist in l for item in sublist]
```

It's a fairly confusing list comprehension, but it *does* work.